PropTypes added. modal windows done. sprint-1/step-2

mcrainbow opened the pull request.

mcrainbow committed:
PropTypes added. modal windows done. sprint-1/step-2

gennady-bars commented on react-burger/src/App.tsx:
```diff
@@ -1,22 +1,53 @@
+      try {
+        const response = await fetch('https://norma.nomoreparties.space/api/ingredients');
+
+        if (!response.ok) {
+          throw new Error('Ошибка: ' + response.status);
+        }
+
+        const jsonData = await response.json();
```
Отлично проверяете ответ на `ok`

gennady-bars commented on react-burger/src/App.tsx:
```diff
@@ -1,22 +1,53 @@
+        console.log(jsonData.data);
+
+        setData(jsonData.data);
+      } catch (err) {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on react-burger/src/App.tsx:
```diff
@@ -1,22 +1,53 @@
+    };
+
+    getData();
+  }, []);
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars commented on react-burger/src/components/BurgerConstructor/BurgerConstructorItem/BurgerConstructorItem.jsx:
```diff
@@ -39,8 +39,20 @@ function BurgerConstructorItem({ item, upperBun }) {
 }
 
```
Вам не нужно верстать вручную ингредиент внутри бургера. Для этого есть компонент `ConstructorElement` в яндекс-библиотеке

gennady-bars commented on react-burger/src/components/BurgerConstructor/BurgerConstructor.jsx:
```diff
@@ -1,48 +1,81 @@
+      <section className={`${burgerConstructorStyle.burgerConstructor} pt-25 pl-4 pr-4 pb-10`}>
+        <ul className={`${burgerConstructorStyle.burgerConstructorList}`}>
+          <BurgerConstructorItem item={data[0]} upperBun={true} />
+          <div className={`${burgerConstructorStyle.burgerConstructorListContainer}`}>
+            <BurgerConstructorItem
+              item={data.find((item) => item.name === 'Соус традиционный галактический')}
+            />
+            <BurgerConstructorItem
+              item={data.find((item) => item.name === 'Мясо бессмертных моллюсков Protostomia')}
+            />
+            <BurgerConstructorItem
+              item={data.find((item) => item.name === 'Плоды Фалленианского дерева')}
+            />
+            <BurgerConstructorItem
+              item={data.find((item) => item.name === 'Плоды Фалленианского дерева')}
+            />
+            <BurgerConstructorItem
+              item={data.find((item) => item.name === 'Соус традиционный галактический')}
```
Для отрисовки **однотипной верстки** принято использовать массив и его метод `map`, чтобы не дублировать код. И обязательно нужно добавлять уникальный атрибут `key` при этом.

gennady-bars commented on react-burger/src/components/BurgerConstructor/BurgerConstructor.jsx:
```diff
@@ -1,48 +1,81 @@
+      </section>
+
+      <Modal
+        isOpen={modalIsOpen}
```
Не нужно передавать в пропсы `Modal` показатель открытости `isOpen`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.


Модальное окно будет появляться целиком вот так:


```jsx

      {modalIsOpen && (
        <Modal
          onClose={() => setModalIsOpen(false)}
          modalClassName="pt-30 pb-30"
        >
          <ModalOrderDetails />
        </Modal>
      )}

```

gennady-bars commented on react-burger/src/components/BurgerConstructor/BurgerConstructor.jsx:
```diff
@@ -1,48 +1,81 @@
-  data: PropTypes.arrayOf(PropTypes.object).isRequired,
+  data: PropTypes.arrayOf(
+    PropTypes.shape({
+      _id: PropTypes.string.isRequired,
+      name: PropTypes.string.isRequired,
+      type: PropTypes.string.isRequired,
+      proteins: PropTypes.number.isRequired,
+      fat: PropTypes.number.isRequired,
+      carbohydrates: PropTypes.number.isRequired,
+      calories: PropTypes.number.isRequired,
+      price: PropTypes.number.isRequired,
+      image: PropTypes.string.isRequired,
+      image_mobile: PropTypes.string,
+      image_large: PropTypes.string,
+      __v: PropTypes.number,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on react-burger/src/components/BurgerConstructor/BurgerConstructorItem/BurgerConstructorItem.jsx:
```diff
@@ -39,8 +39,20 @@ function BurgerConstructorItem({ item, upperBun }) {
 BurgerConstructorItem.propTypes = {
-  isBun: PropTypes.bool,
-  item: PropTypes.object.isRequired,
+  item: PropTypes.shape({
+    _id: PropTypes.string.isRequired,
+    name: PropTypes.string.isRequired,
+    type: PropTypes.string.isRequired,
+    proteins: PropTypes.number.isRequired,
+    fat: PropTypes.number.isRequired,
+    carbohydrates: PropTypes.number.isRequired,
+    calories: PropTypes.number.isRequired,
+    price: PropTypes.number.isRequired,
+    image: PropTypes.string.isRequired,
+    image_mobile: PropTypes.string,
+    image_large: PropTypes.string,
+    __v: PropTypes.number,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on react-burger/src/components/BurgerIngredients/BurgerIngredients.jsx:
```diff
@@ -51,7 +43,22 @@ function BurgerIngredients({ data }) {
 BurgerIngredients.propTypes = {
-  data: PropTypes.arrayOf(PropTypes.object).isRequired,
+  data: PropTypes.arrayOf(
+    PropTypes.shape({
+      _id: PropTypes.string.isRequired,
+      name: PropTypes.string.isRequired,
+      type: PropTypes.string.isRequired,
+      proteins: PropTypes.number.isRequired,
+      fat: PropTypes.number.isRequired,
+      carbohydrates: PropTypes.number.isRequired,
+      calories: PropTypes.number.isRequired,
+      price: PropTypes.number.isRequired,
+      image: PropTypes.string.isRequired,
+      image_mobile: PropTypes.string,
+      image_large: PropTypes.string,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on react-burger/src/components/BurgerIngredients/BurgerIngredientsItem/BurgerIngredientsItem.jsx:
```diff
@@ -1,20 +1,47 @@
+
+BurgerIngredientsItem.propTypes = {
+  item: PropTypes.shape({
+    _id: PropTypes.string.isRequired,
+    name: PropTypes.string.isRequired,
+    type: PropTypes.string.isRequired,
+    proteins: PropTypes.number.isRequired,
+    fat: PropTypes.number.isRequired,
+    carbohydrates: PropTypes.number.isRequired,
+    calories: PropTypes.number.isRequired,
+    price: PropTypes.number.isRequired,
+    image: PropTypes.string.isRequired,
+    image_mobile: PropTypes.string,
+    image_large: PropTypes.string,
+    __v: PropTypes.number,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on react-burger/src/components/Modal/ModalIngredientDetails/ModalIngredientDetails.jsx:
```diff
@@ -0,0 +1,51 @@
+    image: PropTypes.string.isRequired,
+    image_mobile: PropTypes.string,
+    image_large: PropTypes.string.isRequired,
+    __v: PropTypes.number,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on react-burger/src/components/BurgerIngredients/BurgerIngredientsItem/BurgerIngredientsItem.jsx:
```diff
@@ -1,20 +1,47 @@
+        </div>
+      </li>
+
+      <Modal onClose={() => setIsModalOpen(false)} isOpen={isModalOpen}>
```
* Не нужно передавать в пропсы `Modal` показатель открытости `isOpen`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.

gennady-bars replied:
**МОЖНО ЛУЧШЕ**

 Лучше не делать в **каждом** ингредиенте по модальному окну. Нужно было сделать одно окно для деталей в ` BurgerIngredients` и передавать в каждый ингредиент функцию **клика**, в которой бы назначались данные для показа в окне. Это поможет в спринте, где **роутинг** будете изучать и реализовывать


У Вас сейчас 16 попапов одновременно находится в верстке

gennady-bars commented on react-burger/src/components/Modal/ModalIngredientDetails/ModalIngredientDetails.jsx:
```diff
@@ -0,0 +1,51 @@
+import PropTypes from 'prop-types';
+import modalIngredientDetailsStyles from './ModalIngredientDetails.module.css';
+
+export default function ModalIngredientDetails({ item }) {
```
`ModalIngredientDetails` по заданию называется просто `IngredientDetails`. Это детали ингредиента, которые можно вставить в модальное окно, а можно на страницу (это через 2 спринта будет)

gennady-bars commented on react-burger/src/components/Modal/ModalOrderDetails/ModalOrderDetails.jsx:
```diff
@@ -0,0 +1,20 @@
+import modalOrderDetailsStyles from './ModalOrderDetails.module.css';
+import doneImage from '../../../images/done.svg';
+
+export default function ModalOrderDetails() {
```
* `ModalOrderDetails` по заданию называется просто `OrderDetails`. Это детали заказа

gennady-bars commented on react-burger/src/components/Modal/Modal.jsx:
```diff
@@ -0,0 +1,45 @@
+      if (e.key === 'Escape') onClose();
+    };
+    window.addEventListener('keydown', onEsc);
+    return () => window.removeEventListener('keydown', onEsc);
```
Правильно удаляется обработчик

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:


* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз


но есть некоторые недочеты:



* Верхняя и нижняя булки должны **отличаться** по форме. Это делается с помощью атрибутов `type`
* Вам не нужно верстать вручную ингредиент внутри бургера. Для этого есть компонент `ConstructorElement` в яндекс-библиотеке
* Для отрисовки **однотипной верстки** принято использовать массив и его метод `map`, чтобы не дублировать код. И обязательно нужно добавлять уникальный атрибут `key` при этом.
* Не нужно передавать в пропсы `Modal` показатель открытости `isOpen`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.
* Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно
* `ModalIngredientDetails` по заданию называется просто `IngredientDetails`. Это детали ингредиента, которые можно вставить в модальное окно, а можно на страницу (это через 2 спринта будет)
* `ModalOrderDetails` по заданию называется просто `OrderDetails`. Это детали заказа

mcrainbow committed:
Refactor: rename modal components and centralize PropTypes

- Rename ModalIngredientDetails to IngredientDetails
- Rename ModalOrderDetails to OrderDetails
- Remove isOpen prop from Modal component
- Create IngredientType in utils/types.js
- Replace duplicated PropTypes with IngredientType

gennady-bars commented on react-burger/src/components/BurgerConstructor/BurgerConstructor.jsx:
```diff
@@ -1,36 +1,72 @@
-            <BurgerConstructorItem
-              item={data.find((item) => item.name === 'Соус традиционный галактический')}
-            />
+            {ingredients.map((ing) => (
+              <ConstructorElement
+                type="undefind"
+                isLocked={false}
+                text={ing.name}
+                price={200}
```
Еще нужно добавлять `DragIcon` для ингредиентов между булками

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
